python/ray/ml/train/integrations/horovod/horovod_trainer.py

```diff
@@ -0,0 +1,239 @@
+from typing import Dict, Callable, Optional, Union
+
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.ml.trainer import GenDataset
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+
+
+from ray.ml.train.data_parallel_trainer import DataParallelTrainer
+from ray.train.horovod import HorovodConfig
+
+
+class HorovodTrainer(DataParallelTrainer):
+    """A Trainer for data parallel Horovod training.
+
+    This Trainer runs the function ``train_loop_per_worker`` on multiple Ray
+    Actors. These actors already have the necessary Horovod process group already
+    configured for distributed Horovod training.
+
+    The ``train_loop_per_worker`` function is expected to take in either 0 or 1
+    arguments:
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            ...
+
+    .. code-block:: python
+
+        def train_loop_per_worker(config: Dict):
+            ...
+
+    If ``train_loop_per_worker`` accepts an argument, then
+    ``train_loop_config`` will be passed in as the argument. This is useful if you
+    want to tune the values in ``train_loop_config`` as hyperparameters.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards that can then be accessed by ``ray.train.get_dataset_shard("train")`` inside
+    ``train_loop_per_worker``. All the other datasets will not be split and
+    ``ray.train.get_dataset_shard(...)`` will return the the entire Dataset.
+
+    Inside the ``train_loop_per_worker`` function, you can use any of the
+    :ref:`Ray Train function utils <train-api-func-utils>`.
+
+    .. code-block:: python
+
+        def train_loop_per_worker():
+            # Report intermediate results for callbacks or logging.
+            train.report(...)
+
+            # Checkpoints the provided args as restorable state.
+            train.save_checkpoint(...)
+
+            # Returns dict of last saved checkpoint.
+            train.load_checkpoint()
+
+            # Returns the Ray Dataset shard for the given key.
+            train.get_dataset_shard("my_dataset")
+
+            # Returns the total number of workers executing training.
+            train.get_world_size()
+
+            # Returns the rank of this worker.
+            train.get_world_rank()
+
+            # Returns the rank of the worker on the current node.
+            train.get_local_rank()
+
+    You could use ``TensorflowPredictor`` or ``TorchPredictor`` in conjunction with
+    HorovodTrainer. You must save it under the "model" kwarg in
+    ``train.save_checkpoint()``.
+
+    Example:
+
+    .. code-block:: python
+
+        class Net(nn.Module):
```

Comment:
can we use a simpler example for the docstring 🙂

I think we can just copy the one for TorchTrainer, except add in the hvd.init(), hvd.DistributedOptimizer, etc. lines.

Reply:
Yes, good point.
I updated with a simple linear example instead.